src/transformers/models/modernbert/modular_modernbert.py

```diff
@@ -0,0 +1,1515 @@
+
+        Args:
+            hidden_states: (total_nnz, dim)
+            position_ids: (total_nnz,)
+            attention_mask: (batch, max_seqlen)
+            cu_seqlens: (batch + 1,)
+            max_seqlen: int
+        """
+        attn_outputs = self.attn(
+            self.attn_norm(hidden_states),
+            position_ids=position_ids,
+            cu_seqlens=cu_seqlens,
+            max_seqlen=max_seqlen,
+            attention_mask=attention_mask,
+            output_attentions=output_attentions,
+        )
+        hidden_states = hidden_states + attn_outputs[0]
+        hidden_states = hidden_states + self.compiled_mlp(hidden_states)
+
+        return (hidden_states,) + attn_outputs[1:]  # add attentions if outputted
+
+
+class ModernBertPredictionHead(nn.Module):
+    def __init__(self, config: ModernBertConfig):
+        super().__init__()
+        self.config = config
+        self.dense = nn.Linear(config.hidden_size, config.hidden_size, config.classifier_bias)
+        self.act = ACT2FN[config.classifier_activation] if config.classifier_activation else nn.Identity()
+        self.norm = (
+            nn.LayerNorm(config.hidden_size, eps=config.norm_eps, bias=config.norm_bias)
+            if config.classifier_norm
+            else nn.Identity()
+        )
+
+    def _init_weights(self, reset_params: bool = False):
+        if reset_params:
+            self.norm.reset_parameters()
+        _init_modernbert_weights(self.config, self.dense, module_type=ModernBertModuleType.in_module)
+
+    def reset_parameters(self):
+        self._init_weights(reset_params=True)
+
+    def forward(self, hidden_states: torch.Tensor) -> torch.Tensor:
+        return self.norm(self.act(self.dense(hidden_states)))
+
+
+class ModernBertPoolingHead(nn.Module):
+    def __init__(self, config: ModernBertConfig):
+        super().__init__()
+        self.config = config
+        self.dense = nn.Linear(config.hidden_size, config.hidden_size, config.classifier_bias)
+        self.act = ACT2FN[config.classifier_activation] if config.classifier_activation else nn.Identity()
+        self.norm = (
+            nn.LayerNorm(config.hidden_size, eps=config.norm_eps, bias=config.norm_bias)
+            if config.classifier_norm
+            else nn.Identity()
+        )
+        self.drop = torch.nn.Dropout(config.classifier_dropout) if config.classifier_dropout > 0 else nn.Identity()
+        self.pooling_type = ModernBertPoolingType(config.classifier_pooling)
+
+    def forward(self, hidden_states: torch.Tensor, pool: Optional[bool] = True) -> torch.Tensor:
+        if pool:
+            if self.pooling_type == ModernBertPoolingType.cls:
+                output = hidden_states[:, 0]
+            elif self.pooling_type == ModernBertPoolingType.mean:
+                output = hidden_states.mean(dim=1)
+            elif self.pooling_type == ModernBertPoolingType.max:
+                output = hidden_states.max(dim=1)[0]
```

Comment:
same here, we'd rather avoid this, even more in the forward pass. 
Either we have 3 different pooling heads, or you just define 3 functions like for attention! 
The best is to just have the 1 that is used in the released checkpoint

Reply:
I think I'd vote for letting the possibility to choose, we evaluated with CLS pooling because it was the default (and standard across encoders), so the default will most likely be CLS, but mean pooling exhibited strong performance, so letting the choice is a nice feature imho.
(The released checkpoints are not CLS/mean pooling trained, only MLM).

I think 3 pooling heads is fine, but given the philosophy of FlexBert, maybe POOL2FN is more appropriate.

Reply:
Separated into 2 defined functions, much like for Attention, in b3b4028e826d14b623bc6c35f3541ab51a67b234. We're keeping MEAN and CLS as they're both common: CLS is the usual default in HF, but MEAN works a smidge better for some of our tests.
The default is CLS for now.